File: bindings/java/src/main/java/com/github/humancognition/hce/HceCodec.java
package com.github.tmddn3070.hce;

import com.sun.jna.*;
import java.util.Arrays;
import java.util.List;

public final class HceCodec implements AutoCloseable {

    public static final int UNIVERSAL = 0, EU = 1, EN = 2, NUMERIC = 3;
    public static final int SEALED = 0, OPEN = 1, PLAIN = 2;
    public static final int CASE_LOWER = 0, CASE_UPPER = 1;
    public static final int GRAN_SECOND = 0, GRAN_MINUTE = 1, GRAN_HOUR = 2,
                            GRAN_DAY = 3, GRAN_WEEK = 4, GRAN_MONTH = 5;

    interface Lib extends Library {
        Lib INSTANCE = Native.load("hce_ffi", Lib.class);

        Pointer hce_new(byte[] key, long keyLen, byte level, byte mode);
        void hce_destroy(Pointer h);
        void hce_with_bit_width(Pointer h, int bits);
        void hce_with_domain_modulus(Pointer h, long hi, long lo);
        void hce_with_cipher_kind(Pointer h, byte kind, byte[] key, long keyLen);
        void hce_with_case(Pointer h, byte caseVal);
        void hce_with_check_syllables(Pointer h, long n);
        void hce_with_separator(Pointer h, byte sep);
        void hce_with_chunk_none(Pointer h);
        void hce_with_chunk_fixed(Pointer h, long charSize);
        void hce_with_chunk_pattern(Pointer h, long[] pattern, long count);
        void hce_with_timestamp_config(Pointer h, long epochMs, byte granularity);
        Result.ByValue hce_encode(Pointer h, byte[] data, long len);
        Result.ByValue hce_decode(Pointer h, byte[] input, long len);
        RecoveryResult.ByValue hce_recover(Pointer h, byte[] input, long len);
        void hce_free_result(Result.ByValue r);
        void hce_free_recovery_result(RecoveryResult.ByValue r);
    }

    public static class Result extends Structure {
        public byte ok; public Pointer data; public long len; public int errCode;
        @Override protected List<String> getFieldOrder() { return Arrays.asList("ok","data","len","errCode"); }
        public static class ByValue extends Result implements Structure.ByValue {}
        byte[] toBytes() { if (ok == 0 || data == null || len == 0) return new byte[0]; return data.getByteArray(0, (int) len); }
    }

    public static class RecoveryResult extends Structure {
        public byte ok; public Pointer corrected; public long correctedLen; public long candidateCount; public int errCode;
        @Override protected List<String> getFieldOrder() { return Arrays.asList("ok","corrected","correctedLen","candidateCount","errCode"); }
        public static class ByValue extends RecoveryResult implements Structure.ByValue {}
    }

    private final Pointer inner;

    public HceCodec(byte[] key, int level, int mode) {
        if (mode != PLAIN && (key == null || key.length == 0)) throw new IllegalArgumentException("key required");
        byte[] k = key != null ? key : new byte[0];
        var h = Lib.INSTANCE.hce_new(k, k.length, (byte) level, (byte) mode);
        if (h == null) throw new RuntimeException("hce_new failed");
        this.inner = h;
    }

    public String encode(byte[] data) {
        if (data == null || data.length == 0) throw new IllegalArgumentException("data required");
        var r = Lib.INSTANCE.hce_encode(inner, data, data.length);
        if (r.ok == 0) throw new RuntimeException("encode failed: " + r.errCode);
        var out = r.toBytes(); Lib.INSTANCE.hce_free_result(r); return new String(out);
    }
    public byte[] decode(String input) {
        var in = input.getBytes();
        var r = Lib.INSTANCE.hce_decode(inner, in, in.length);
        if (r.ok == 0) throw new RuntimeException("decode failed: " + r.errCode);
        var out = r.toBytes(); Lib.INSTANCE.hce_free_result(r); return out;
    }
    public byte[] recover(String input) {
        var in = input.getBytes();
        var r = Lib.INSTANCE.hce_recover(inner, in, in.length);
        try {
            if (r.ok == 0) throw new RuntimeException("recover failed: " + r.errCode);
            if (r.correctedLen > 0 && r.corrected != null) {
                byte[] out = new byte[(int) r.correctedLen];
                r.corrected.read(0, out, 0, out.length);
                return out;
            }
            return decode(input);
        } finally { Lib.INSTANCE.hce_free_recovery_result(r); }
    }

    public HceCodec withBitWidth(int bits) { Lib.INSTANCE.hce_with_bit_width(inner, bits); return this; }
    public HceCodec withModulus(long modulus) {
        Lib.INSTANCE.hce_with_domain_modulus(inner, modulus >>> 32, modulus & 0xFFFFFFFFL); return this; }
    public HceCodec withCipherFeistel(byte[] key) {
        byte[] k = key != null ? key : new byte[0];
        Lib.INSTANCE.hce_with_cipher_kind(inner, (byte)0, k, k.length); return this; }
    public HceCodec withCipherShuffle() { Lib.INSTANCE.hce_with_cipher_kind(inner, (byte)1, null, 0); return this; }
    public HceCodec withCase(int caseVal) { Lib.INSTANCE.hce_with_case(inner, (byte) caseVal); return this; }
    public HceCodec withCheckSyllables(int n) { Lib.INSTANCE.hce_with_check_syllables(inner, n); return this; }
    public HceCodec withSeparator(char sep) { Lib.INSTANCE.hce_with_separator(inner, (byte) sep); return this; }
    public HceCodec withChunkNone() { Lib.INSTANCE.hce_with_chunk_none(inner); return this; }
    public HceCodec withChunkFixed(int charSize) { Lib.INSTANCE.hce_with_chunk_fixed(inner, charSize); return this; }
    public HceCodec withChunkPattern(int... pattern) {
        var p = new long[pattern.length]; for (int i = 0; i < pattern.length; i++) p[i] = pattern[i];
        Lib.INSTANCE.hce_with_chunk_pattern(inner, p, p.length); return this; }
    public HceCodec withTimestampConfig(long epochMs, int gran) {
        Lib.INSTANCE.hce_with_timestamp_config(inner, epochMs, (byte) gran); return this; }

    @Override public void close() { if (inner != null) Lib.INSTANCE.hce_destroy(inner); }
}

File: bindings/swift/Sources/Hce/HceCodec.swift
import Foundation

@_silgen_name("hce_new")
func hce_new(_ key: UnsafePointer<UInt8>?, _ keyLen: Int, _ level: UInt8, _ mode: UInt8) -> OpaquePointer?

@_silgen_name("hce_destroy")
func hce_destroy(_ h: OpaquePointer?)

@_silgen_name("hce_with_bit_width")
func hce_with_bit_width(_ h: OpaquePointer?, _ bits: UInt32)

@_silgen_name("hce_with_domain_modulus")
func hce_with_domain_modulus(_ h: OpaquePointer?, _ hi: UInt64, _ lo: UInt64)

@_silgen_name("hce_with_cipher_kind")
func hce_with_cipher_kind(_ h: OpaquePointer?, _ kind: UInt8, _ key: UnsafePointer<UInt8>?, _ keyLen: Int)

@_silgen_name("hce_with_case")
func hce_with_case(_ h: OpaquePointer?, _ caseVal: UInt8)

@_silgen_name("hce_with_check_syllables")
func hce_with_check_syllables(_ h: OpaquePointer?, _ n: Int)

@_silgen_name("hce_with_separator")
func hce_with_separator(_ h: OpaquePointer?, _ sep: UInt8)

@_silgen_name("hce_with_chunk_none")
func hce_with_chunk_none(_ h: OpaquePointer?)

@_silgen_name("hce_with_chunk_fixed")
func hce_with_chunk_fixed(_ h: OpaquePointer?, _ charSize: Int)

@_silgen_name("hce_with_chunk_pattern")
func hce_with_chunk_pattern(_ h: OpaquePointer?, _ pattern: UnsafePointer<Int>?, _ count: Int)

@_silgen_name("hce_with_timestamp_config")
func hce_with_timestamp_config(_ h: OpaquePointer?, _ epochMs: Int64, _ gran: UInt8)

@_silgen_name("hce_encode")
func hce_encode(_ h: OpaquePointer?, _ data: UnsafePointer<UInt8>?, _ len: Int) -> HceResult

@_silgen_name("hce_decode")
func hce_decode(_ h: OpaquePointer?, _ input: UnsafePointer<UInt8>?, _ len: Int) -> HceResult

@_silgen_name("hce_recover")
func hce_recover(_ h: OpaquePointer?, _ input: UnsafePointer<UInt8>?, _ len: Int) -> RecoveryResult

@_silgen_name("hce_free_result")
func hce_free_result(_ r: HceResult)

@_silgen_name("hce_free_recovery_result")
func hce_free_recovery_result(_ r: RecoveryResult)

struct HceResult {
    var ok: Bool
    var data: UnsafePointer<UInt8>?
    var len: Int
    var errCode: Int32
}

struct RecoveryResult {
    var ok: Bool
    var corrected: UnsafePointer<UInt8>?
    var correctedLen: Int
    var candidateCount: Int
    var errCode: Int32
}

public final class HceCodec {
    private var ptr: OpaquePointer?

    public enum Level: UInt8 { case universal = 0, eu = 1, en = 2, numeric = 3 }
    public enum Mode: UInt8  { case sealed = 0, open = 1, plain = 2 }
    public enum CipherCase: UInt8 { case lower = 0, upper = 1 }
    public enum Granularity: UInt8 { case second = 0, minute = 1, hour = 2, day = 3, week = 4, month = 5 }

    public init?(key: Data?, level: Level, mode: Mode) {
        if mode != .plain && (key == nil || key!.isEmpty) { return nil }
        let kp = key?.withUnsafeBytes { $0.baseAddress?.assumingMemoryBound(to: UInt8.self) }
        guard let h = hce_new(kp, key?.count ?? 0, level.rawValue, mode.rawValue) else { return nil }
        ptr = h
    }

    deinit { if let p = ptr { hce_destroy(p) } }

    @discardableResult public func withBitWidth(_ bits: UInt32) -> Self { hce_with_bit_width(ptr, bits); return self }
    @discardableResult public func withModulus(_ m: UInt64) -> Self { hce_with_domain_modulus(ptr, UInt64(m >> 32), UInt64(m & 0xFFFF_FFFF)); return self }
    @discardableResult public func withCipherFeistel(_ key: Data?) -> Self {
        let kp = key?.withUnsafeBytes { $0.baseAddress?.assumingMemoryBound(to: UInt8.self) }
        hce_with_cipher_kind(ptr, 0, kp, key?.count ?? 0); return self
    }
    @discardableResult public func withCipherShuffle() -> Self { hce_with_cipher_kind(ptr, 1, nil, 0); return self }
    @discardableResult public func withCase(_ c: CipherCase) -> Self { hce_with_case(ptr, c.rawValue); return self }
    @discardableResult public func withCheckSyllables(_ n: Int) -> Self { hce_with_check_syllables(ptr, n); return self }
    @discardableResult public func withSeparator(_ ch: Character) -> Self { hce_with_separator(ptr, UInt8(ch.asciiValue ?? 45)); return self }
    @discardableResult public func withChunkNone() -> Self { hce_with_chunk_none(ptr); return self }
    @discardableResult public func withChunkFixed(_ n: Int) -> Self { hce_with_chunk_fixed(ptr, n); return self }
    @discardableResult public func withChunkPattern(_ pattern: [Int]) -> Self {
        let p = pattern; p.withUnsafeBufferPointer { hce_with_chunk_pattern(ptr, $0.baseAddress, $0.count) }; return self
    }
    @discardableResult public func withTimestampConfig(_ epochMs: Int64, _ gran: Granularity) -> Self { hce_with_timestamp_config(ptr, epochMs, gran.rawValue); return self }

    public func encode(_ data: Data) -> String? {
        let r = data.withUnsafeBytes { b in hce_encode(ptr, b.baseAddress?.assumingMemoryBound(to: UInt8.self), data.count) }
        guard r.ok, let d = r.data else { return nil }
        defer { hce_free_result(r) }
        return String(bytes: UnsafeBufferPointer(start: d, count: r.len), encoding: .utf8)
    }

    public func decode(_ input: String) -> Data? {
        let bytes = [UInt8](input.utf8)
        let r = bytes.withUnsafeBufferPointer { hce_decode(ptr, $0.baseAddress, input.utf8.count) }
        guard r.ok, let d = r.data else { return nil }
        defer { hce_free_result(r) }
        return Data(bytes: d, count: r.len)
    }

    public func recover(_ input: String) -> Data? {
        let bytes = [UInt8](input.utf8)
        let r = bytes.withUnsafeBufferPointer { hce_recover(ptr, $0.baseAddress, input.utf8.count) }
        defer { hce_free_recovery_result(r) }
        guard r.ok else { return nil }
        if r.correctedLen > 0, let c = r.corrected { return Data(bytes: c, count: r.correctedLen) }
        return decode(input)
    }
}
